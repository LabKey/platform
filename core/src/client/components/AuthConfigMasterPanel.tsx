import React, { PureComponent } from 'react';

import { Panel, DropdownButton, MenuItem, Tab, Tabs } from 'react-bootstrap';

import { LabelHelpTip } from '@labkey/components';

import DragAndDropPane from './DragAndDropPane';
import AuthRow from './AuthRow';
import DynamicConfigurationModal from './DynamicConfigurationModal';

interface ViewOnlyAuthConfigRowsProps {
    data: AuthConfig[];
    providers?: AuthConfigProvider[];
}

class ViewOnlyAuthConfigRows extends PureComponent<ViewOnlyAuthConfigRowsProps> {
    render() {
        const { providers, data } = this.props;

        return (
            <div>
                {data &&
                data.map(item => (
                    <AuthRow
                        {...item}
                        key={item.configuration}
                        canEdit={false}
                        draggable={false}
                        modalType={providers && { ...providers[item.provider] }}
                    />
                ))}
            </div>
        );
    }
}

interface PrimaryTabProps {
    canEdit: boolean;
    addNewPrimaryDropdown: JSX.Element[];
    loginFormTipText: string;
    primaryTabLoginForm: JSX.Element;
    SSOTipText: string;
    primaryTabSSO: JSX.Element;
}

class PrimaryTab extends PureComponent<PrimaryTabProps> {
    render() {
        const {
            addNewPrimaryDropdown,
            loginFormTipText,
            primaryTabLoginForm,
            SSOTipText,
            primaryTabSSO,
            canEdit
        } = this.props;

        return(
            <>
                <div className="configurations__auth-tab">
                    {canEdit && (
                        <div className="configurations__dropdown">
                            <DropdownButton
                                id="primary-configurations-dropdown"
                                title="Add New Primary Configuration">
                                {addNewPrimaryDropdown}
                            </DropdownButton>
                        </div>
                    )}

                    <div className="configurations__config-section-title">
                        <span className="bold-text"> Login Form Configurations </span>
                        <LabelHelpTip
                            title="Tip"
                            body={() => {
                                return <div> {loginFormTipText} </div>;
                            }}
                        />
                    </div>

                    {primaryTabLoginForm}
                </div>

                <div>
                    <div className="configurations__config-section-title">
                        <span className="bold-text"> Single Sign On Configurations </span>
                        <LabelHelpTip
                            title="Tip"
                            body={() => {
                                return <div> {SSOTipText} </div>;
                            }}
                        />
                    </div>

                    {primaryTabSSO}
                </div>
            </>
        );
    }
}

interface SecondaryTabProps {
    canEdit: boolean;
    addNewSecondaryDropdown: JSX.Element[];
    secondaryTabDuo: JSX.Element;
}

class SecondaryTab extends PureComponent<SecondaryTabProps> {
    render() {
        const {canEdit, addNewSecondaryDropdown, secondaryTabDuo} = this.props;

        return(
            <div className="configurations__auth-tab">
                {canEdit && (
                    <div className="configurations__dropdown">
                        <DropdownButton
                            id="secondary-configurations-dropdown"
                            title="Add New Secondary Configuration">
                            {addNewSecondaryDropdown}
                        </DropdownButton>
                    </div>
                )}

                {secondaryTabDuo}
            </div>
        );
    }
}

interface Props {
    formConfigurations?: AuthConfig[];
    ssoConfigurations?: AuthConfig[];
    secondaryConfigurations?: AuthConfig[];
    primaryProviders?: AuthConfigProvider[];
    secondaryProviders?: AuthConfigProvider[];
    helpLink?: string;
    canEdit?: boolean;
    isDragDisabled?: boolean;
    actions?: Actions;
}

interface State {
    primaryModalOpen?: boolean;
    secondaryModalOpen?: boolean;
    addModalType?: string | null;
}

export default class AuthConfigMasterPanel extends PureComponent<Props, State> {
    constructor(props) {
        super(props);
        this.state = {
            primaryModalOpen: false,
            secondaryModalOpen: false,
            addModalType: null,
        };
    }

    onToggleModal = (toggled: string): void => {
        this.setState(() => ({
            [toggled]: !this.state[toggled],
        }));
    };

    determineconfigType = (addModalType: string): string => {
        let authType;
        if (addModalType in this.props.primaryProviders) {
            const authInfo = this.props.primaryProviders[addModalType];

            if (authInfo.sso) {
                authType = 'ssoConfigurations';
            } else {
                authType = 'formConfigurations';
            }
        } else {
            authType = 'secondaryConfigurations';
        }
        return authType;
    };

    render() {
        const {
            primaryProviders,
            secondaryProviders,
            ssoConfigurations,
            formConfigurations,
            secondaryConfigurations,
            canEdit,
        } = this.props;

        const SSOTipText = 'These configurations let LabKey users authenticate against an external service such as a SAML identify provider or a CAS server. LabKey will render SSO logos in the header and on the login page in the order that the configurations are listed below.';
        const loginFormTipText = "These configurations make use of LabKey's login page to collect credentials and authenticate against either hashed credentials stored in the LabKey database or external LDAP servers. LabKey will attempt authenticating against each configuration in the order they are listed below.";
        const authenticationDocsLink = this.props.helpLink;

        const addNewPrimaryDropdown =
            primaryProviders &&
            Object.keys(primaryProviders).map(authOption => (
                <MenuItem
                    key={authOption}
                    onClick={() => this.setState({ primaryModalOpen: true, addModalType: authOption })}>
                    {authOption} : {primaryProviders[authOption].description}
                </MenuItem>
            ));

        const addNewSecondaryDropdown =
            secondaryProviders &&
            Object.keys(secondaryProviders).map(authOption => (
                <MenuItem
                    key={authOption}
                    onClick={() => this.setState({ secondaryModalOpen: true, addModalType: authOption })}>
                    {authOption} : {secondaryProviders[authOption].description}
                </MenuItem>
            ));

        const dbAuth = formConfigurations && formConfigurations.slice(-1)[0];

        const primaryTabLoginForm =
            formConfigurations && canEdit ? (
                <div>
                    <DragAndDropPane
                        configType="formConfigurations"
                        authConfigs={formConfigurations.slice(0, -1)} // Database config is excluded from DragAndDrop
                        providers={primaryProviders}
                        canEdit={canEdit}
                        isDragDisabled={this.props.isDragDisabled}
                        actions={this.props.actions}
                    />

                    <AuthRow
                        canEdit={canEdit}
                        draggable={false}
                        description={dbAuth.description}
                        provider={dbAuth.provider}
                        enabled={dbAuth.enabled}
                        toggleModalOpen={this.props.actions.toggleModalOpen}
                    />
                </div>
            ) : (
                <ViewOnlyAuthConfigRows data={formConfigurations} providers={primaryProviders} />
            );

        const primaryTabSSO =
            ssoConfigurations && canEdit ? (
                <DragAndDropPane
                    configType="ssoConfigurations"
                    authConfigs={ssoConfigurations}
                    providers={primaryProviders}
                    canEdit={canEdit}
                    isDragDisabled={this.props.isDragDisabled}
                    actions={this.props.actions}
                />
            ) : (
                <ViewOnlyAuthConfigRows data={ssoConfigurations} providers={primaryProviders} />
            );

        const secondaryTabDuo =
            secondaryConfigurations && canEdit ? (
                <DragAndDropPane
                    configType="secondaryConfigurations"
                    authConfigs={secondaryConfigurations}
                    providers={secondaryProviders}
                    canEdit={canEdit}
                    isDragDisabled={this.props.isDragDisabled}
                    actions={this.props.actions}
                />
            ) : (
                <ViewOnlyAuthConfigRows data={secondaryConfigurations} providers={secondaryProviders} />
            );

        const dataBaseModal = (this.state.primaryModalOpen || this.state.secondaryModalOpen) &&
                <DynamicConfigurationModal
                        modalType={
                            this.state.primaryModalOpen
                                ? this.props.primaryProviders[this.state.addModalType]
                                : this.props.secondaryProviders[this.state.addModalType]
                        }
                        configType={this.determineconfigType(this.state.addModalType)}
                        description={this.state.addModalType + ' Configuration'}
                        enabled={true}
                        canEdit={canEdit}
                        title={'New ' + this.state.addModalType + ' Configuration'}
                        provider={this.state.addModalType}
                        updateAuthRowsAfterSave={this.props.actions.updateAuthRowsAfterSave}
                        closeModal={() => {
                            this.onToggleModal(
                                this.state.primaryModalOpen ? 'primaryModalOpen' : 'secondaryModalOpen'
                            );
                            this.props.actions.toggleModalOpen(false);
                        }}
                />;

        return (
            <Panel>
                <Panel.Heading>
                    <span className="bold-text"> Configurations </span>
                </Panel.Heading>
                <Panel.Body>
                    <a className="configurations__help-link" href={authenticationDocsLink}>
                        Get help with authentication
                    </a>

                    {dataBaseModal}

                    <Tabs defaultActiveKey={1} id="tab-panel">
                        <Tab eventKey={1} title="Primary">
                            <PrimaryTab
                                canEdit={canEdit}
                                addNewPrimaryDropdown={addNewPrimaryDropdown}
                                loginFormTipText={loginFormTipText}
                                primaryTabLoginForm={primaryTabLoginForm}
                                SSOTipText={SSOTipText}
                                primaryTabSSO={primaryTabSSO}
                            />
                        </Tab>

                        <Tab eventKey={2} title="Secondary">
                            <SecondaryTab
                                canEdit={canEdit}
                                addNewSecondaryDropdown={addNewSecondaryDropdown}
                                secondaryTabDuo={secondaryTabDuo}
                            />
                        </Tab>
                    </Tabs>
                </Panel.Body>
            </Panel>
        );
    }
}
