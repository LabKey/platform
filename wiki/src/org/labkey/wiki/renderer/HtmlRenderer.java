/*
 * Copyright (c) 2008-2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.labkey.wiki.renderer;

import org.apache.commons.lang3.StringUtils;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.labkey.api.attachments.Attachment;
import org.labkey.api.util.PageFlowUtil;
import org.labkey.api.util.TidyUtil;
import org.labkey.api.view.template.ClientDependency;
import org.labkey.api.wiki.WikiRenderer;
import org.labkey.api.wiki.FormattedHtml;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * User: Tamra Myers
 * Date: Aug 16, 2006
 * Time: 12:33:37 PM
 */
public class HtmlRenderer implements WikiRenderer
{
    private final String _hrefPrefix;
    private final String _attachPrefix;
    private final Map<String, String> _nameTitleMap;
    private final Map<String, Attachment> _attachments;

    private static Map<String, SubstitutionHandler> _substitutionHandlers = new HashMap<>();

    static
    {
        // Add SubstitutionHandlers for each ${labkey.<type>()}
        _substitutionHandlers.put("webPart", new WebPartSubstitutionHandler());
        _substitutionHandlers.put("dependency", new ClientDependencySubstitutionHandler());
    }


    public HtmlRenderer(String hrefPrefix, String attachPrefix, Map<String, String> nameTitleMap, @Nullable Collection<? extends Attachment> attachments)
    {
        _hrefPrefix = hrefPrefix;
        _attachPrefix = attachPrefix;
        _nameTitleMap = nameTitleMap == null ? new HashMap<>() : nameTitleMap;
        _attachments = new HashMap<>();

        if (null != attachments)
            for (Attachment a : attachments)
                _attachments.put(a.getName(), a);
    }
    

    public FormattedHtml format(String text)
    {
        LinkedList<String> errors = new LinkedList<>();
        if (text == null)
            return new FormattedHtml("");

        // Remove degenerate comments (e.g. "<!-->") as Tidy does not handle them properly
        text = text.replaceAll("<!--*>|<!>", "");

        FormattedHtml formattedHtml = handleLabKeySubstitutions(text);
        boolean volatilePage = formattedHtml.isVolatile();
        LinkedHashSet<ClientDependency> cds = new LinkedHashSet<>(formattedHtml.getClientDependencies());

        Document doc = TidyUtil.convertHtmlToDocument("<html><body>" + StringUtils.trimToEmpty(formattedHtml.getHtml()) + "</body></html>", false, errors);
        if (!errors.isEmpty() || doc == null)
        {
            StringBuilder innerHtml = new StringBuilder("<div class=\"labkey-error\"><b>An exception occurred while generating the HTML.  Please correct this content.</b></div><br>The error message was: ");
            if (!errors.isEmpty())
            {
                for (String error : errors)
                    innerHtml.append(PageFlowUtil.filter(error)).append("<hr>");
            }
            return new FormattedHtml(innerHtml.toString(), false, cds);
        }

        Set<String> wikiDependencies = new HashSet<>();

        // process A and IMG
        NodeList nl = doc.getElementsByTagName("a");
        Map<Element, String> linkExceptions = new HashMap<>();
        for (int i=0 ; i<nl.getLength() ; i++)
        {
            Element a = (Element)nl.item(i);
            try
            {
                String href = PageFlowUtil.decode(a.getAttribute("href"));

                Attachment at = _attachments.get(href);
                if (null != at)
                {
                    a.setAttribute("href", _attachPrefix + PageFlowUtil.encode(at.getName()));
                    wikiDependencies.add(at.getName());
                    continue;
                }

                if (href.isBlank() || href.startsWith("#"))
                    continue;

                String wikiName = StringUtils.substringBefore(href, "#");
                String title = _nameTitleMap.get(wikiName);
                if (null != title)
                {
                    // UNDONE: why is l.getName() null???
                    //a.setAttribute("href", _hrefPrefix + PageFlowUtil.encode(.getName()));
                    a.setAttribute("href", _hrefPrefix + PageFlowUtil.encode(href));
                    wikiDependencies.add(wikiName);
                    continue;
                }

                // if this doesn't look like a url, then link to it as a wiki page
                if (StringUtils.containsNone(href, "?:/.&"))
                {
                    a.setAttribute("href", _hrefPrefix + PageFlowUtil.encode(href));
                    wikiDependencies.add(wikiName);
                }
            }
            catch (IllegalArgumentException e)
            {
                // Issue 12160: there was an issue with decoding the href for the link, so store the exception text and link index for later
                linkExceptions.put(a, e.getMessage());
            }
        }

        nl = doc.getElementsByTagName("img");
        for (int i=0 ; i<nl.getLength() ; i++)
        {
            Element img = (Element)nl.item(i);
            String src = img.getAttribute("src");
            src = PageFlowUtil.decode(src);
            Attachment at = _attachments.get(src);
            if (null != at)
            {
                img.setAttribute("src", _attachPrefix + PageFlowUtil.encode(at.getName()));
                wikiDependencies.add(at.getName());
            }
        }

        // back to html
        StringBuilder innerHtml;

        try
        {
            innerHtml = new StringBuilder();

            //look for style elements, as tidy moves them to the head section
            NodeList styleNodes = doc.getElementsByTagName("style");
            for (int idx = 0; idx < styleNodes.getLength(); ++idx)
            {
                innerHtml.append(PageFlowUtil.convertNodeToHtml(styleNodes.item(idx)));
                innerHtml.append('\n');
            }

            Node bodyNode = doc.getElementsByTagName("body").item(0);
            // Uncomment the below to debug bodyNode contents
            // inspectNode(bodyNode, 0);

            String bodyHtml = PageFlowUtil.convertNodeToHtml(bodyNode);

            // Issue 12160: if there were any link decoding exceptions, replace the link with the error message
            if (linkExceptions.size() > 0)
            {
                for (Map.Entry<Element, String> link : linkExceptions.entrySet())
                {
                    String linkHtml = "<span class='labkey-error'>" + link.getValue() + "</span>";
                    bodyHtml = bodyHtml.replace(PageFlowUtil.convertNodeToHtml(link.getKey()), linkHtml);
                }
            }

            innerHtml.append(bodyHtml, "<body>".length(), bodyHtml.length()-"</body>".length());
        }
        catch (Exception e)
        {
            innerHtml = new StringBuilder("<div class=\"labkey-error\"><b>An exception occurred while generating the HTML.  Please correct this content.</b></div><br>The error message was: ");
            innerHtml.append(e.getMessage()); 
            volatilePage = false;
        }

        return new FormattedHtml(innerHtml.toString(), volatilePage, wikiDependencies, Collections.emptySet(), cds);
    }


    // ${labkey.<type>(<any_stream of characters>)}
    // Pattern.DOTALL allows the parameter list to span multiple lines
    private static final Pattern _substitutionPattern = Pattern.compile("\\$\\{labkey\\.(\\w+)\\((.*?)\\)\\}", Pattern.DOTALL);

    // <any word>='<any value>', allowing whitespace before, after, and in-between
    private static final Pattern _paramPattern = Pattern.compile("\\s*(\\w+)\\s*=\\s*'(.*)'\\s*");

    private FormattedHtml handleLabKeySubstitutions(String text)
    {
        if (text == null)
            return new FormattedHtml("");
        
        // Find all substitution templates embedded in wiki text that have the form ${labkey.<type>(<any_stream of characters>)}.
        Matcher webPartMatcher = _substitutionPattern.matcher(text);

        // If we find none, return immediately
        if (!webPartMatcher.find())
            return new FormattedHtml(text);

        List<Definition> definitions = new ArrayList<>(10);
        Map<Definition, List<String>> wikiErrors = new HashMap<>();
        do
        {
            List<String> paramErrors = new ArrayList<>();
            String substitutionType = webPartMatcher.group(1);          // type
            String params = webPartMatcher.group(2).replace(",", "");
            // Parse the parameters with the symbols in parseWith, they can be used in any order
            // as long as they are the same symbol starts and completes a parameter value
            List<String> paramList = new ArrayList<>();
            paramList.add(params);
            String[] parseWith = { "&#39;", "'" };
            for (String parser : parseWith)
            {
                List<String> paramListTemp = new ArrayList<>();
                for (String paramSection : paramList)
                {
                    String[] paramSplit = paramSection.split(parser);
                    for (int i = 0; i < paramSplit.length; i+=2)
                    {
                        paramListTemp.add(paramSplit[i] + (paramSplit.length > i + 1 ? "'" + paramSplit[i+1]  + "'": ""));
                    }
                }
                paramList = paramListTemp;
            }

            Map<String, String> paramMap = new HashMap<>(10);
            for (String param : paramList)
            {
                Matcher paramMatcher = _paramPattern.matcher(param);

                if (paramMatcher.matches())
                {
                    if (paramMap.containsKey(paramMatcher.group(1)))
                        paramErrors.add(param.trim() + ", there are multiple parameters with this name");
                    else
                        paramMap.put(paramMatcher.group(1), paramMatcher.group(2));
                }
                else if (param.trim().length() > 0)
                    paramErrors.add(param.trim());
            }

            // Stick new definition at beginning of list -- we want to replace them in reverse order
            Definition definition = new Definition(substitutionType, webPartMatcher.start(), webPartMatcher.end(), paramMap);
            definitions.add(0, definition);
            wikiErrors.put(definition, paramErrors);
        }
        while(webPartMatcher.find());

        StringBuilder sb = new StringBuilder(text);
        boolean volatilePage = false;
        LinkedHashSet<ClientDependency> cds = new LinkedHashSet<>();

        // Get the corresponding substitution handler for each type and replace template with substitution
        for (Definition definition : definitions)
        {
            SubstitutionHandler handler = _substitutionHandlers.get(definition.getType());
            FormattedHtml substitution;

            if (null != handler)
                substitution = handler.getSubstitution(definition.getParams());
            else
                substitution = new FormattedHtml("<br><font class='error' color='red'>Error: unknown type, \"labkey." + definition.getType() + "\"</font>");

            sb.replace(definition.getStart(), definition.getEnd(), substitution.getHtml());

            if (substitution.isVolatile())
                volatilePage = true;

            cds.addAll(substitution.getClientDependencies());

            List<String> paramErrors = wikiErrors.get(definition);
            if (paramErrors.size() > 0)
            {
                String errorHTML = "<br>";
                for (String error : paramErrors)
                    errorHTML = errorHTML.concat("<font class='error' color='red'>Error with parameter " +
                            error + " in " + definition.getType() + "</font><br><br>");
                sb.insert(definition.getStart() + substitution.getHtml().length(), errorHTML);
            }
        }

        return new FormattedHtml(sb.toString(), volatilePage, cds);
    }


    private static class Definition
    {
        private final String _type;
        private final int _start;
        private final int _end;
        private final Map<String, String> _params;

        private Definition(String type, int start, int end, Map<String, String> params)
        {
            _type = type;
            _start = start;
            _end = end;
            _params = params;
        }

        public String getType()
        {
            return _type;
        }

        private int getEnd()
        {
            return _end;
        }

        private Map<String, String> getParams()
        {
            return _params;
        }

        private int getStart()
        {
            return _start;
        }
    }


    interface SubstitutionHandler
    {
        @NotNull FormattedHtml getSubstitution(Map<String, String> params);
    }
}
