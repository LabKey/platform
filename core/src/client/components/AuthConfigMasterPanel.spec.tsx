import React from 'react';
import AuthConfigMasterPanel from './AuthConfigMasterPanel';
import renderer from 'react-test-renderer';

describe("<AuthConfigMasterPanel/>", () => {
    test("Editable mode", () => {
        let basicFn = () => {};
        let actionFns = {"name": basicFn};
        const component =
            <AuthConfigMasterPanel
                canEdit={true}
                actionFunctions={actionFns}
            />;

        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("View-only mode", () => {
        let basicFn = () => {};
        let actionFns = {"name": basicFn};
        const component =
            <AuthConfigMasterPanel
                canEdit={false}
                actionFunctions={actionFns}
            />;

        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });
});