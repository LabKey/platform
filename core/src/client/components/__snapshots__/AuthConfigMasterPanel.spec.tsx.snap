// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthConfigMasterPanel/> Editable mode 1`] = `
<div
  className="panel panel-default"
>
  <div
    className="panel-heading"
  >
    <span
      className="bold-text"
    >
       Configurations 
    </span>
  </div>
  <div
    className="panel-body"
  >
    <a
      className="configurations__help-link"
    >
      Get help with authentication
    </a>
    <div
      id="tab-panel"
    >
      <ul
        className="nav nav-tabs"
        role="tablist"
      >
        <li
          className="active"
          role="presentation"
        >
          <a
            aria-controls="tab-panel-pane-1"
            aria-selected={true}
            href="#"
            id="tab-panel-tab-1"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="tab"
          >
            Primary
          </a>
        </li>
        <li
          className=""
          role="presentation"
        >
          <a
            aria-controls="tab-panel-pane-2"
            aria-selected={false}
            href="#"
            id="tab-panel-tab-2"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="tab"
            tabIndex={-1}
          >
            Secondary
          </a>
        </li>
      </ul>
      <div
        className="tab-content"
      >
        <div
          aria-hidden={false}
          aria-labelledby="tab-panel-tab-1"
          className="fade tab-pane active in"
          id="tab-panel-pane-1"
          role="tabpanel"
        >
          <div
            className="configurations__auth-tab"
          >
            <div
              className="configurations__dropdown"
            >
              <div
                className="dropdown btn-group"
              >
                <button
                  aria-expanded={false}
                  aria-haspopup={true}
                  className="dropdown-toggle btn btn-default"
                  disabled={false}
                  id="primary-configurations-dropdown"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                  type="button"
                >
                  Add New Primary Configuration
                   
                  <span
                    className="caret"
                  />
                </button>
                <ul
                  aria-labelledby="primary-configurations-dropdown"
                  className="dropdown-menu"
                  role="menu"
                />
              </div>
            </div>
            <div
              className="configurations__config-section-title"
            >
              <span
                className="bold-text"
              >
                 Login Form Configurations 
              </span>
              <span
                className="label-help-target"
                onMouseOut={[Function]}
                onMouseOver={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-question-circle fa-w-16 fa-1x label-help-icon"
                  data-icon="question-circle"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </span>
            </div>
            <div />
          </div>
          <div>
            <div
              className="configurations__config-section-title"
            >
              <span
                className="bold-text"
              >
                 Single Sign On Configurations 
              </span>
              <span
                className="label-help-target"
                onMouseOut={[Function]}
                onMouseOver={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-question-circle fa-w-16 fa-1x label-help-icon"
                  data-icon="question-circle"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </span>
            </div>
            <div />
          </div>
        </div>
        <div
          aria-hidden={true}
          aria-labelledby="tab-panel-tab-2"
          className="fade tab-pane"
          id="tab-panel-pane-2"
          role="tabpanel"
        >
          <div
            className="configurations__auth-tab"
          >
            <div
              className="configurations__dropdown"
            >
              <div
                className="dropdown btn-group"
              >
                <button
                  aria-expanded={false}
                  aria-haspopup={true}
                  className="dropdown-toggle btn btn-default"
                  disabled={false}
                  id="secondary-configurations-dropdown"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                  type="button"
                >
                  Add New Secondary Configuration
                   
                  <span
                    className="caret"
                  />
                </button>
                <ul
                  aria-labelledby="secondary-configurations-dropdown"
                  className="dropdown-menu"
                  role="menu"
                />
              </div>
            </div>
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<AuthConfigMasterPanel/> View-only mode 1`] = `
<div
  className="panel panel-default"
>
  <div
    className="panel-heading"
  >
    <span
      className="bold-text"
    >
       Configurations 
    </span>
  </div>
  <div
    className="panel-body"
  >
    <a
      className="configurations__help-link"
    >
      Get help with authentication
    </a>
    <div
      id="tab-panel"
    >
      <ul
        className="nav nav-tabs"
        role="tablist"
      >
        <li
          className="active"
          role="presentation"
        >
          <a
            aria-controls="tab-panel-pane-1"
            aria-selected={true}
            href="#"
            id="tab-panel-tab-1"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="tab"
          >
            Primary
          </a>
        </li>
        <li
          className=""
          role="presentation"
        >
          <a
            aria-controls="tab-panel-pane-2"
            aria-selected={false}
            href="#"
            id="tab-panel-tab-2"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="tab"
            tabIndex={-1}
          >
            Secondary
          </a>
        </li>
      </ul>
      <div
        className="tab-content"
      >
        <div
          aria-hidden={false}
          aria-labelledby="tab-panel-tab-1"
          className="fade tab-pane active in"
          id="tab-panel-pane-1"
          role="tabpanel"
        >
          <div
            className="configurations__auth-tab"
          >
            <div
              className="configurations__config-section-title"
            >
              <span
                className="bold-text"
              >
                 Login Form Configurations 
              </span>
              <span
                className="label-help-target"
                onMouseOut={[Function]}
                onMouseOver={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-question-circle fa-w-16 fa-1x label-help-icon"
                  data-icon="question-circle"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </span>
            </div>
            <div />
          </div>
          <div>
            <div
              className="configurations__config-section-title"
            >
              <span
                className="bold-text"
              >
                 Single Sign On Configurations 
              </span>
              <span
                className="label-help-target"
                onMouseOut={[Function]}
                onMouseOver={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-question-circle fa-w-16 fa-1x label-help-icon"
                  data-icon="question-circle"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </span>
            </div>
            <div />
          </div>
        </div>
        <div
          aria-hidden={true}
          aria-labelledby="tab-panel-tab-2"
          className="fade tab-pane"
          id="tab-panel-pane-2"
          role="tabpanel"
        >
          <div
            className="configurations__auth-tab"
          >
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
